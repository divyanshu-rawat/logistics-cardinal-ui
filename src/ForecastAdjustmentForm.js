import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { FormControl, FormGroup, Button } from 'react-bootstrap';

import ForecastAdjustmentPreviewTable from './ForecastAdjustmentPreviewTable';
import { StyledSelectFull } from './ForecastStyledSelect';
import reasonOptions from './forecastReasonOptions.json';

const FormDiv = styled('div')`
  .form-section {
    padding: 10px 0;
    .forecast-form-buttons,
    .forecast-form,
    .forecast-form-adjustments {
      display: flex;
      justify-content: space-between;
      > div {
        width: 49%;
      }
    }
    .forecast-form-adjustments {
      display: flex;
      justify-content: space-between;
      > div:first-of-type {
        padding-right: 10px;
      }
    }
    .forecast-form-reasons {
      display: block;
      > div {
        width: 100%;
      }
    }
    .dates {
      color: ${(props) => props.theme.inactive};
      line-height: 39px;
      padding: 8px;
      display: flex;
      justify-content: space-between;
      .time {
        min-width: 130px;
        text-align: right;
      }
    }
    label[for='adjustment'],
    label[for='comment'] {
      display: none;
    }
    .right-side {
      float: right;
    }
    .react-bs-container-body table {
      border-top: 1px solid #40464f;
      border-left: 1px solid #40464f;
      border-right: 1px solid #40464f;
      tr {
        border-bottom: 1px solid #40464f;
      }
    }
    textarea,
    input {
      padding: 5px 10px;
      line-height: 25px;
    }
    input.form-control {
      height: 32px;
    }
    .checkbox {
      line-height: 32px;
      margin: 0;
      input {
        margin-top: 10px;
      }
    }
    .forecast-form-buttons button {
      width: 160px;
    }
  }
`;

export function ForecastAdjustmentForm({
  filters,
  updateAdjustment,
  adjustment,
  onSave,
  onCancel,
}) {
  return (
    <FormDiv>
      <div className="form-section">
        <div className="forecast-form">
          <div>
            <div className="dates">
              <span>
                {adjustment.selectedDays.map((dayData) => (
                  <span key={`${dayData.dataObj.week}${dayData.dataObj.day}`}>
                    {dayData.date.format('ddd, MMM Do')};{' '}
                  </span>
                ))}
              </span>
              <span className="time">
                <span>
                  {filters.startTime} - {filters.endTime}
                </span>
              </span>
            </div>
            <div className="forecast-form-reasons form-group">
              <StyledSelectFull
                searchable={false}
                clearable={false}
                name="reason"
                title="Reason"
                placeholder="Reason"
                value={adjustment.reason}
                options={reasonOptions}
                onChange={(option) =>
                  updateAdjustment({ reason: option.value })
                }
              />
            </div>
            <FormGroup controlId="reasonDetail">
              <FormControl
                name="reasonDetail"
                title="Reason Detail"
                placeholder="Reason Detail"
                componentClass="textarea"
                value={adjustment.reasonDetail}
                onChange={(event) =>
                  updateAdjustment({ reasonDetail: event.target.value })
                }
              />
            </FormGroup>
            <div className="forecast-form-adjustments">
              <FormGroup controlId="adjustment">
                <FormControl
                  name="adjustment"
                  title="Adjustment %"
                  placeholder="Adjustment %"
                  type="number"
                  value={adjustment.adjustment}
                  onChange={(event) =>
                    updateAdjustment({
                      adjustment: parseInt(event.target.value, 10) || '',
                    })
                  }
                />
              </FormGroup>
              {/* Hide the Fix Adjustment, we will display it again in the future
                  <Checkbox
                    name="fixAdjustment"
                    title="Fix Adjustment"
                    placeholder="Fix Adjustment"
                    checked={adjustment.fixAdjustment}
                    onChange={(event) =>
                      updateAdjustment({ fixAdjustment: event.target.checked })
                    }
                  >
                    Fix Adjustment
                  </Checkbox>
                */}
              <StyledSelectFull
                searchable={false}
                clearable={false}
                name="appliesTo"
                title="Applies to..."
                placeholder="Applies to..."
                multi
                onChange={(options) => updateAdjustment({ appliesTo: options })}
                value={adjustment.appliesTo}
                options={[
                  { label: 'base', value: 'base' },
                  { label: 'd2', value: 'd2' },
                ]}
              />
            </div>
          </div>
          <div>
            <ForecastAdjustmentPreviewTable
              adjustment={adjustment.adjustment || 0}
              filters={filters}
              selectedDays={adjustment.selectedDays}
            />
          </div>
        </div>
      </div>
      <div className="form-section">
        <div className="forecast-form-buttons">
          <div>
            <Button bsSize="small" className="right-side" onClick={onCancel}>
              Cancel
            </Button>
          </div>
          <div>
            <Button bsSize="small" bsStyle="primary" onClick={onSave}>
              Save
            </Button>
          </div>
        </div>
      </div>
    </FormDiv>
  );
}

ForecastAdjustmentForm.propTypes = {
  filters: PropTypes.object.isRequired,
  updateAdjustment: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  adjustment: PropTypes.object.isRequired,
};

export default ForecastAdjustmentForm;
