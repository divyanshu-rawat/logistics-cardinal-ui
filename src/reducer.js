import moment from 'moment-timezone';
import cloneDeep from 'lodash/cloneDeep';
import * as constants from './constants';
import {
  createForecastDataStructure,
  formatZoneList,
  updateSelectedDays,
  updateHistory,
} from './utils';

const momentStartDate = moment()
  .utc()
  .startOf('day')
  .subtract(4, 'weeks')
  .startOf('isoWeek');
const momentEndDate = moment()
  .utc()
  .endOf('day')
  .add(2, 'weeks')
  .endOf('isoWeek');

export const forecastData = {
  charts: {},
  tableValues: [],
  weekDates: {},
  biggestOrdersNumber: 0,
  biggestTotalValue: 0,
  colors: {},
};

export const initialState = {
  originalData: {},
  ...forecastData,
  filters: {
    startDate: momentStartDate.format('YYYY-MM-DD'),
    startWeek: momentStartDate.isoWeek(),
    startTime: '09:00:00',
    startFullDate: momentStartDate.toISOString(),
    endDate: momentEndDate.format('YYYY-MM-DD'),
    endWeek: momentEndDate.isoWeek(),
    endTime: '22:30:00',
    endFullDate: momentEndDate.toISOString(),
    zoneIds: [],
  },
  zones: [],
  adjustment: {
    selectedDays: [],
    adjustment: '',
    fixAdjustment: false,
    reason: '',
    reasonDetail: '',
    appliesTo: ['base'],
    loading: false,
    error: null,
  },
  loading: false,
  error: null,
  history: {
    filters: {
      startTime: '00:00:00',
      endTime: '23:59:59',
      zoneIds: [],
    },
    data: [],
    error: null,
    loading: false,
  },
  forecastNeedsUpdate: false,
};

function reducer(state = initialState, action) {
  switch (action.type) {
    case constants.GET_FORECAST: {
      return {
        ...state,
        loading: true,
        error: null,
      };
    }
    case constants.GET_FORECAST_ERROR: {
      const { error } = action;
      return {
        ...state,
        error,
        loading: false,
      };
    }

    case constants.GET_FORECAST_SUCCESS: {
      const formattedForecast = createForecastDataStructure(
        action.forecast,
        state.filters,
      );
      let updatedSelectedDays;
      // if there are selected days already, they must be updated
      if (state.adjustment.selectedDays.length) {
        updatedSelectedDays = updateSelectedDays(
          formattedForecast.tableValues,
          state.adjustment.selectedDays,
        );
      }
      return {
        ...state,
        loading: false,
        error: null,
        forecastNeedsUpdate: false,
        ...formattedForecast,
        originalData: action.forecast,
        adjustment: {
          ...state.adjustment,
          selectedDays: updatedSelectedDays || state.adjustment.selectedDays,
        },
      };
    }

    case constants.UPDATE_FORECAST_FILTERS: {
      const currentFilters = state.filters;
      const newFilters = { ...currentFilters, ...action.dataObject };

      const formattedForecast = createForecastDataStructure(
        state.originalData,
        newFilters,
      );

      return {
        ...state,
        filters: { ...newFilters },
        ...formattedForecast,
        adjustment: {
          ...state.adjustment,
          selectedDays: [],
        },
      };
    }

    case constants.GET_ZONE_IDS_SUCCESS:
      return {
        ...state,
        loading: false,
        error: null,
        zones: formatZoneList(action.zones),
      };

    case constants.GET_ZONE_IDS_ERROR: {
      const { error } = action;
      return {
        ...state,
        error,
        loading: false,
      };
    }

    case constants.UPDATE_SELECTED_DAYS: {
      const { date, dataObj } = action;
      const selectedDays = cloneDeep(state.adjustment.selectedDays);
      let dayExists = false;

      state.adjustment.selectedDays.some((day, index) => {
        if (day.date.isSame(date)) {
          dayExists = true;
          selectedDays.splice(index, 1);
          return true;
        }

        return false;
      });
      if (!dayExists) {
        selectedDays.push({ date, dataObj });
        selectedDays.sort((a, b) => (a.date.isBefore(b.date) ? -1 : 1));
      }

      return { ...state, adjustment: { ...state.adjustment, selectedDays } };
    }

    case constants.UPDATE_ADJUSTMENT:
      return {
        ...state,
        adjustment: { ...state.adjustment, ...action.dataObj },
      };

    case constants.CANCEL_ADJUSTMENT:
      return {
        ...state,
        adjustment: { ...state.adjustment, selectedDays: [] },
      };

    case constants.UPDATE_HISTORY_FILTERS: {
      const currentFilters = state.history.filters;
      const newFilters = { ...currentFilters, ...action.dataObject };

      return {
        ...state,
        history: {
          ...state.history,
          filters: { ...newFilters },
        },
      };
    }

    case constants.GET_ADJUSTMENTS: {
      return {
        ...state,
        history: {
          ...state.history,
          loading: true,
          error: null,
        },
      };
    }
    case constants.GET_ADJUSTMENTS_ERROR: {
      const { error } = action;
      return {
        ...state,
        history: {
          ...state.history,
          loading: false,
          error,
        },
      };
    }

    case constants.GET_ADJUSTMENTS_SUCCESS: {
      return {
        ...state,
        history: {
          ...state.history,
          data: action.adjustments,
        },
      };
    }

    case constants.POST_ADJUSTMENT: {
      return {
        ...state,
        adjustment: {
          ...state.adjustment,
          loading: true,
          error: null,
        },
      };
    }
    case constants.POST_ADJUSTMENT_SUCCESS: {
      const originalData = {
        ...state.originalData,
        forecasts: action.result.forecasts,
      };
      const formattedForecast = createForecastDataStructure(
        originalData,
        state.filters,
      );

      return {
        ...state,
        adjustment: initialState.adjustment,
        originalData,
        ...formattedForecast,
      };
    }

    case constants.POST_ADJUSTMENT_ERROR: {
      return {
        ...state,
        adjustment: {
          ...state.adjustment,
          loading: false,
          error: action.error,
        },
      };
    }

    case constants.PATCH_ADJUSTMENT_SUCCESS: {
      return {
        ...state,
        history: {
          ...state.history,
          data: updateHistory(state.history.data, action.result),
        },
        forecastNeedsUpdate: true,
      };
    }

    case constants.PATCH_ADJUSTMENT_ERROR: {
      return {
        ...state,
        history: {
          ...state.history,
          loading: false,
          error: action.error,
        },
      };
    }

    default:
      return state;
  }
}

export default reducer;
