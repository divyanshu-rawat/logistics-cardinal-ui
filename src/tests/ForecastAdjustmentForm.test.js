import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { noop } from 'lodash';

import ForecastAdjustmentForm from '../ForecastAdjustmentForm';
import { initialState } from '../reducer';

describe('<ForecastAdjustmentForm />', () => {
  let renderedComponent;

  beforeEach(() => {
    renderedComponent = shallow(
      <ForecastAdjustmentForm
        filters={initialState.filters}
        selectedDays={[]}
        updateAdjustment={noop}
        onSave={noop}
        adjustment={initialState.adjustment}
        onCancel={noop}
      />,
    );
  });

  it('should render the component', () => {
    const tree = toJson(renderedComponent);
    expect(tree).toMatchSnapshot();
  });
});
