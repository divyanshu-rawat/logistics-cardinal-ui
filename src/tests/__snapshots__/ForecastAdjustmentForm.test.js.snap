// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForecastAdjustmentForm /> should render the component 1`] = `
<styled.div>
  <div
    className="form-section"
  >
    <div
      className="forecast-form"
    >
      <div>
        <div
          className="dates"
        >
          <span />
          <span
            className="time"
          >
            <span>
              09:00:00
               - 
              22:30:00
            </span>
          </span>
        </div>
        <div
          className="forecast-form-reasons form-group"
        >
          <Styled(StyledSelect)
            clearable={false}
            name="reason"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "Trend not captured",
                  "value": "trend",
                },
                Object {
                  "label": "Event of public interest",
                  "value": "event",
                },
                Object {
                  "label": "Special Day",
                  "value": "special_day",
                },
                Object {
                  "label": "Marketing Activity",
                  "value": "marketing",
                },
                Object {
                  "label": "Other",
                  "value": "other",
                },
              ]
            }
            placeholder="Reason"
            searchable={false}
            title="Reason"
            value=""
          />
        </div>
        <FormGroup
          bsClass="form-group"
          controlId="reasonDetail"
        >
          <FormControl
            bsClass="form-control"
            componentClass="textarea"
            name="reasonDetail"
            onChange={[Function]}
            placeholder="Reason Detail"
            title="Reason Detail"
            value=""
          />
        </FormGroup>
        <div
          className="forecast-form-adjustments"
        >
          <FormGroup
            bsClass="form-group"
            controlId="adjustment"
          >
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="adjustment"
              onChange={[Function]}
              placeholder="Adjustment %"
              title="Adjustment %"
              type="number"
              value=""
            />
          </FormGroup>
          <Styled(StyledSelect)
            clearable={false}
            multi={true}
            name="appliesTo"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "base",
                  "value": "base",
                },
                Object {
                  "label": "d2",
                  "value": "d2",
                },
              ]
            }
            placeholder="Applies to..."
            searchable={false}
            title="Applies to..."
            value={
              Array [
                "base",
              ]
            }
          />
        </div>
      </div>
      <div>
        <ForecastAdjustmentPreviewTable
          adjustment={0}
          filters={
            Object {
              "endDate": "2018-11-25",
              "endFullDate": "2018-11-25T23:59:59.999Z",
              "endTime": "22:30:00",
              "endWeek": 47,
              "startDate": "2018-10-08",
              "startFullDate": "2018-10-08T00:00:00.000Z",
              "startTime": "09:00:00",
              "startWeek": 41,
              "zoneIds": Array [],
            }
          }
          selectedDays={Array []}
        />
      </div>
    </div>
  </div>
  <div
    className="form-section"
  >
    <div
      className="forecast-form-buttons"
    >
      <div>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="right-side"
          disabled={false}
          onClick={[Function]}
        >
          Cancel
        </Button>
      </div>
      <div>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="primary"
          disabled={false}
          onClick={[Function]}
        >
          Save
        </Button>
      </div>
    </div>
  </div>
</styled.div>
`;
